File: Medium/Python/DynamicProgramming/CoinChange.py
class Solution(object):
    def coinChange(self, coins, amount):
        """
            :type coins: List[int]
            :type amount: int
            :rtype: int
            """
        if not coins:
            if amount == 0:
                return 0
            return -1
        dp = [0] + [float("inf")] * amount
        coins.sort()
        for i in xrange(1, amount + 1):
            for coin in coins:
                if i >= coin:
                    dp[i] = min(dp[i], dp[i - coin] + 1)
        return dp[amount] if dp[amount] <= amount else -1

File: Medium/Java/DynamicProgramming/CoinChange.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        if (coins.length < 0)
            return -1;
        int[] res = new int[amount + 1];
        for (int i = 1; i < res.length; i ++)
            res[i] = Integer.MAX_VALUE - 1;
        Arrays.sort(coins);
        for (int i = 1; i < res.length; i ++)
            for (int coin: coins) {
                if (i >= coin)
                    res[i] = Math.min(res[i], res[i - coin] + 1);
            }
        return res[amount] <= amount? res[amount]: -1;
    }
}
